### ISSUE TITLE
Fix contest finding #4

### ISSUE THREAD

**Issue by baroooo** [mento-protocol/mento-core#407]
### Description

https://github.com/sherlock-audit/2024-02-mento-judging/issues/4

We decided to fix it by passing the history as a storage variable.

We should also add tests to measure the gas cost of the functions that use balanceOf and totalSupply with a high number of locks and various timings such as with a new lock just before a proposal creation, after a proposal creation, in the same block with a castVote etc.

### Acceptance Criteria
- [x] Issue fixed
- [x] Gas cost is reasonable
- [x] PR is opened to fixes branch
- [x] Fix is added to the sherlock fixes review

**Comment by zigtur** [comment]
Hi @baroooo 

On my side, I have tested this fix (setting history as a storage variable instead of memory, which requires a view function instead of pure).

I tested multiple exploits and didn't succeed to trigger high gas consumption even by registering 100K+ locks in a single block (which is highly unlikely to happen).

The fix seems to be efficient. Adding tests for this specific edgecase is a really good reaction though!
